indexer/src/handler.rs

```diff
@@ -0,0 +1,220 @@
+
+use backoff::ExponentialBackoff;
+use dashmap::DashMap;
+use futures::{sink::SinkExt, stream::StreamExt};
+use holaplex_hub_nfts_solana_core::{
+    db::Connection,
+    proto::{
+        solana_nft_events::Event::UpdateMintOwner, MintOwnershipUpdate, SolanaNftEventKey,
+        SolanaNftEvents,
+    },
+    CollectionMint,
+};
+use hub_core::{prelude::*, producer::Producer, tokio::task};
+use solana_client::rpc_client::RpcClient;
+use solana_program::program_pack::Pack;
+use solana_sdk::{pubkey::Pubkey, signature::Signature};
+use spl_token::{instruction::TokenInstruction, state::Account};
+use yellowstone_grpc_client::GeyserGrpcClientError;
+use yellowstone_grpc_proto::{
+    prelude::{subscribe_update::UpdateOneof, SubscribeUpdate, SubscribeUpdateTransaction},
+    tonic::Status,
+};
+
+use crate::{Args, GeyserGrpcConnector};
+
+#[derive(Clone)]
+pub struct MessageHandler {
+    db: Connection,
+    dashmap: DashMap<u64, Vec<SubscribeUpdateTransaction>>,
+    rpc: Arc<RpcClient>,
+    connector: GeyserGrpcConnector,
+    producer: Producer<SolanaNftEvents>,
+}
+
+impl MessageHandler {
+    pub async fn new(args: Args, producer: Producer<SolanaNftEvents>) -> Result<Self> {
+        let Args {
+            endpoint,
+            x_token,
+            solana_endpoint,
+            db,
+        } = args;
+
+        let db = Connection::new(db)
+            .await
+            .context("failed to get database connection")?;
+        let dashmap: DashMap<u64, Vec<SubscribeUpdateTransaction>> = DashMap::new();
+        let rpc = Arc::new(RpcClient::new(solana_endpoint));
+        let connector = GeyserGrpcConnector::new(endpoint, x_token);
+
+        Ok(Self {
+            db,
+            dashmap,
+            rpc,
+            connector,
+            producer,
+        })
+    }
+
+    pub async fn run(&self) -> Result<()> {
+        let request = GeyserGrpcConnector::build_request();
+        let (mut subscribe_tx, mut stream) = self.connector.subscribe().await?;
+
+        loop {
+            let request = request.clone();
+
+            subscribe_tx
+                .send(request)
+                .await
+                .map_err(GeyserGrpcClientError::SubscribeSendError)?;
+
+            while let Some(message) = stream.next().await {
+                self.handle_message(message).await?;
+            }
+        }
+    }
+
+    async fn handle_message(&self, message: Result<SubscribeUpdate, Status>) -> Result<()> {
+        match message {
+            Ok(msg) => match msg.update_oneof {
+                Some(UpdateOneof::Transaction(tx)) => {
+                    self.dashmap.entry(tx.slot).or_insert(Vec::new()).push(tx);
+                },
+                Some(UpdateOneof::Slot(slot)) => {
+                    if let Some((_, transactions)) = self.dashmap.remove(&slot.slot) {
+                        for tx in transactions {
+                            task::spawn(self.clone().process_transaction(tx));
```

Comment:
We use tokio yield_now in our main function for redpanda message handling is this something that should be done here @ray-kast ? Can you explain when we should yield?